ReactWindows/ReactNative/Common/AccessibilityHelper.cs

```diff
@@ -0,0 +1,292 @@
+using ReactNative.Bridge;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Automation;
+using Windows.UI.Xaml.Automation.Peers;
+
+namespace ReactNative.Common
+{
+    /// <summary>
+    /// Accessibility helper methods.
+    /// </summary>
+    public static class AccessibilityHelper
+    {
+        #region ImportantForAccessibility - Provides support for ImportantForAccessibility property implementation.
+
+        /// <summary>
+        /// Initializes the value of AccessibilityView for <paramref name="uiElement"/> and its children,
+        /// based on the ancestor <paramref name="parentUIElement"/> settings. Must be called
+        /// when an element is initially added to the visual tree.
+        /// </summary>
+        /// <param name="parentUIElement">The element where <paramref name="uiElement"/> is added as child.</param>
+        /// <param name="uiElement">The element being added to the visual tree.</param>
+        public static void InitImportantForAccessibility(UIElement parentUIElement, UIElement uiElement)
+        {
+            DispatcherHelpers.RunOnDispatcher(() =>
```

Comment:
>DispatcherHelpers.RunOnDispatcher(() => [](start = 12, length = 39)

`InitImportantForAccessibility` starts executing on the dispatcher so my assumption is that the purpose of this is we need to wait for other operations to finish running on the dispatcher before we run our logic.

Is there some other way to ensure this runs in the correct order relative to other operations?

Patterns like this concern me because I'm not confident that this operation will actually run in the desired order relative to other operations. Maybe it's possible this will still run too soon or maybe it'll run too late. Maybe this view will have been removed from the tree by the time our code runs.

Maybe see if Radu has any alternative ideas about how to solve this.

Reply:
I talked to Radu about this and he suggested exploring ViewParentManager::UpdateExtraData. However I found that it is called much more often than just on adding children and if that is the case there will be perf impact of traversing the tree unnecessarily. I understand the concern of implying order but the code doesn't rely on it - it would be dangerous to rely on a particular order and also it can break easily. This is more like an extra assurance. Apparently the operation that adds children works bottom up (Radu confirmed) and posting on the UI thread will run the code after children has been added so that parent chain is established. But even if that is not the case and the order changes there will an operation where the very top node of the chain will have the children added. At that point there is logic to update the the top node children and that logic will not be dependent on parent (since this is top node). So after all, the status will be correct.

---
In reply to: [165803827](https://github.com/Microsoft/react-native-windows/pull/1595#discussion_r165803827) [](ancestors = 165803827)

Reply:
I think we should consider 2 things:
  - Closing with the XAML team on when the text's automation tree will be ready to traverse
  - Based on this, we might be able to find or introduce a hook into react-native-windows for running this importantForAccessibility code.

---
In reply to: [165824073](https://github.com/Microsoft/react-native-windows/pull/1595#discussion_r165824073) [](ancestors = 165824073,165803827)